### PR TITLE
Cache magic.link's token 

### LINKED ISSUE #2501: Cache magic.link's token
We re-request a token from magic.link with every request. This is really inefficient because the token is valid for 15 minutes. Magic.link doesn't return an expiration time, so we should cache the token for at least ~10 minutes.

image

Here are the steps to fix this:

1. Grab and cache the token (in memory) over here: https://github.com/redwoodjs/redwood/blob/ef0d09350ef53de8251e0970f690444c1d1a218f/packages/auth/src/authClients/magicLink.ts#L23
This function is called for each graphql-request, so we can use it to invalidate that cached token.
2. Store a timestamp for when the token should expire (now + 10 minutes).
3. With each request check if the timestamp has expired, if it has re-request a new token and recycle.

### PR DESCRIPTION
**Pull request by simoncrypta**
This cache magic.link's token for 10min 
fix #2501

### CODE REVIEW

File: packages/auth/src/authClients/magicLink.ts [hunk 1]
```diff
@@ -8,6 +8,8 @@ export interface AuthClientMagicLink extends AuthClient {
 }
 
 export const magicLink = (client: MagicLink): AuthClientMagicLink => {
+  let token: string | null
+  let expireTime: number
```
Comment (peterp) [inline packages/auth/src/authClients/magicLink.ts:12]: I would set the default to now, then you can remove the extra check on line 24.
```suggestion
  let expireTime = Date.now(): number
```

File: packages/auth/src/authClients/magicLink.ts [hunk 2]
```diff
@@ -18,7 +20,15 @@ export const magicLink = (client: MagicLink): AuthClientMagicLink => {
     },
     signup: async ({ email, showUI }) =>
       await client.auth.loginWithMagicLink({ email, showUI }),
-    getToken: async () => await client.user.getIdToken(),
+    getToken: async () => {
+      if (!token || !expireTime || Date.now() <= expireTime) {
```
Comment (peterp) [inline packages/auth/src/authClients/magicLink.ts:24]:
```suggestion
      if (!token || Date.now() <= expireTime) {
```

**Review by peterp — CHANGES_REQUESTED** [review]
Thanks for doing this! Just a small suggestion to remove the extra check.

**Comment by peterp** [comment]
@Burnsy Thought you would be interested in seeing the solution for this

File: packages/auth/src/authClients/magicLink.ts (hunk 1 above)
Comment (simoncrypta) [reply packages/auth/src/authClients/magicLink.ts:12]: Good point ! 😄

**Comment by Burnsy** [comment]
This looks great will try it in my own app soon 👍🏻

**Comment by Burnsy** [comment]
@simoncrypta  just tried this in my own app and the code works great! LGTM @peterp

**Review by peterp — APPROVED** [review]

**Pull request merged**